File: 3. Longest Substring Without Repeating Characters/3. Longest Substring Without Repeating Characters.cs
using System;
using System.Collections.Generic;

public class Solution {
    public int LengthOfLongestSubstring(string s) {
        Dictionary<char, int> map = new Dictionary<char, int>();
        int maxLen = 0, start = 0;
        
        for (int end = 0; end < s.Length; end++) {
            if (map.ContainsKey(s[end])) {
                start = Math.Max(map[s[end]] + 1, start);
            }
            map[s[end]] = end;
            maxLen = Math.Max(maxLen, end - start + 1);
        }
        
        return maxLen;
    }
    
    // Uncomment the main method if testing locally. Remove this when submitting to avoid multiple entry points.
    /*
    public static void Main(string[] args) {
        string s = "abcabcbb";
        Solution solution = new Solution();
        Console.WriteLine("Length of longest substring: " + solution.LengthOfLongestSubstring(s));
    }
    */
}

File: 3. Longest Substring Without Repeating Characters/3. Longest Substring Without Repeating Characters.dart
class Solution {
  int lengthOfLongestSubstring(String s) {
    Map<String, int> map = {};
    int maxLen = 0, start = 0;

    for (int end = 0; end < s.length; end++) {
      String current = s[end];
      if (map.containsKey(current)) {
        start = (map[current]! + 1 > start) ? map[current]! + 1 : start;
      }
      map[current] = end;
      maxLen = (end - start + 1 > maxLen) ? end - start + 1 : maxLen;
    }

    return maxLen;
  }
}